x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java

```diff
@@ -171,48 +174,64 @@ private void normalStopDatafeed(Task task, StopDatafeedAction.Request request, A
 
         ActionListener<StopDatafeedAction.Response> finalListener = ActionListener.wrap(
                 r -> waitForDatafeedStopped(allDataFeedsToWaitFor, request, r, listener),
-                listener::onFailure);
+                e -> {
+                    if (ExceptionsHelper.unwrapCause(e) instanceof FailedNodeException) {
+                        // A node has dropped out of the cluster since we started executing the requests.
+                        // Since stopping an already stopped datafeed is not an error we can try again.
+                        // The datafeeds that were running on the node that dropped out of the cluster
+                        // will just have their persistent tasks cancelled.  Datafeeds that were stopped
+                        // by the previous attempt will be noops in the subsequent attempt.
+                        doExecute(task, request, listener);
+                    } else {
+                        listener.onFailure(e);
+                    }
+                });
 
         super.doExecute(task, request, finalListener);
     }
 
     private void forceStopDatafeed(final StopDatafeedAction.Request request, final ActionListener<StopDatafeedAction.Response> listener,
```

Comment:
Sure, that's done in the second commit.

While I was doing this I noticed that although we have a `starting` state we never use it.  So I also added a TODO about using that for 8.0.  (Looking back through the history, `starting` was added in 5.5, but couldn't be used until 6.x because 5.4 wouldn't understand it.  Now we're beyond that we can use it, but we should only do this in a major version just in case somebody is relying on `stopped`, `started` and `stopping` being the only 3 states that exist.

Reply:
I'm not sure we need to use it. I recall adding it so we had it available without BWC in case we needed it. But surely, if it improves things we can definitely use it.

Reply:
I think not using it creates a potential race condition where if you start and stop a datafeed in very quick succession then the stop will be ignored.  It seems that none of our current tests do this.